Issue/1482 step3 handle discovery error changes

anitaa1990 opened the pull request:
Partially fixes #1482. This PR completes the third step of the Jetpack sign in with self credentials as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1482).

### Changes:
- ~~If a discovery error takes place for a site,~~
- Further testing revealed that some errors such as XMLRPC_FORBIDDEN, XMLRPC_BLOCKED and MISSING_XMLRPC_METHOD errors can occur even with Jetpack available on the site. Furthermore, the `CONNECT_SITE_INFO` API returns that Jetpack does not exist, even when Jetpack is available and connected on the site. In order to address this issue, the logic for discovery errors has being modified as follows:
  - For HTTP Auth error: HTTP Auth is not supported by Jetpack so we can safely assume that this error would only occur if Jetpack is not connected on a site. So we redirect the user to the Jetpack required screen. There is a possibility that user installs Jetpack while on this screen and clicks on `Try again`. In order to avoid confusion and displaying No jetpack available when it is, we are using the `CONNECT_SITE_INFO` API to verify if Jetpack is installed/active/connected. 
    - If Jetpack is connected, redirect them back to the site credentials screen.
    - If Jetpack is still not connected, display an error snackbar asking them to install Jetpack.
  - All other errors that can occur: Regardless of whether Jetpack is available or note, redirect them to the discovery error screen. 
    - Added a new fragment called `LoginDiscoveryErrorFragment` that is displayed when a discovery error takes place for a site that has jetpack available.
     - Clicking on `Sign in with WordPress.com` redirects  to the email screen.
     - Clicking on `Read our troubleshooting tips` redirects to this [link](https://jetpack.com/support/getting-started-with-jetpack/troubleshooting-tips/) inside a webview.
     - Clicking on `Try again` redirects to the site credentials screen.
- Ported over discovery initiation from the `LoginSiteAddressFragment` to the `LoginBaseDiscoveryFragment`.

### Notes:
- The are TODO's added to the PR to add events to the button clicks. This will take place in another PR.
- Note that this PR is to be merged to a feature branch.

### Screenshots
#### Logging into a site with HTTP AUTH required when Jetpack is NOT available:
(LEFT: When jetpack is still not installed and user clicks on Try AGAIN button)
(RIGHT: When jetpack is installed when user is on the screen and clicks on TRY AGAIN button)
<img width="300" src="https://user-images.githubusercontent.com/22608780/67929599-b12a5400-fbe3-11e9-91fe-4ef425c5c734.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/67929603-b25b8100-fbe3-11e9-96d3-b67c79b079d1.gif">

#### Logging into a site with XMLRPC missing when Jetpack is available:
<img width="300" src="https://user-images.githubusercontent.com/22608780/67929521-88a25a00-fbe3-11e9-9c0f-eb899e47221f.gif">

#### Logging into a site with missing rsd tags:
<img width="300" src="https://user-images.githubusercontent.com/22608780/67952094-bce13f00-fc12-11e9-84d0-41367c700472.gif">

#### Logging into a site with SSL Certificate error :
<img width="300" src="https://user-images.githubusercontent.com/22608780/68001780-97444c00-fc8b-11e9-815d-5529e9d04e0c.gif">

#### Logging into a site with the XML-RPC request is blocked :
<img width="300" src="https://user-images.githubusercontent.com/22608780/68087189-3da17480-fe79-11e9-9ff0-63ed38c863ec.gif">

#### Different discovery error messages:
(LEFT: HTTP AUTH required . RIGHT: SSL certificate needed)
<img width="300" src="https://user-images.githubusercontent.com/22608780/67869344-4b8e8700-fb53-11e9-8651-c6d63cd82e06.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/67869420-68c35580-fb53-11e9-85e3-47b234d84066.png">

(LEFT: XMLRPC errors . RIGHT: All other generic errors)
<img width="300" src="https://user-images.githubusercontent.com/22608780/67869850-03239900-fb54-11e9-944e-49ff7ebfa5a0.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/67869983-382feb80-fb54-11e9-9b50-801d5ae925d4.png">

### Testing:
##### Scenario I: Login to a site where the XMLRPC file is renamed.  
(This scenario is applicable to the following discovery errors: `MISSING_XMLRPC_METHOD`, `XMLRPC_BLOCKED`, `XMLRPC_FORBIDDEN`, `GENERIC_ERROR`, `NO_SITE_FOUND`)
- Login to anitaamurthy.com in `wp-admin`.
- Install, activate and connect Jetpack.
- Open the app and enter the above site address and click on `Next`.
- Verify that you are redirected to the Sign in with Jetpack screen (even though Jetpack is installed on the site). This is because the XMLRPC file is renamed for this site and cannot be found so even if Jetpack is available, the API response returns false for the fields `	hasJetpack: true`, `isJetpackActive: false`, `isJetpackConnected: false`.
- Click on Sign in in the app.
- Click on `Login with site credentials` and enter username + password.
- Verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access the XMLRPC file on your site. You will need to reach out to your host to resolve this.`.
- Verify that clicking on `Try again` redirects to the Username + password screen.
- Verify that clicking on `Read our troubleshooting tips` redirects to a webview from this link.
- Verify that clicking on `Help` opens the help page.
- Verify that clicking in `Sign in with Wordpress.com` redirects to the Email screen and you are able to login successfully to the app using wp.com email.

##### Scenario II: Login to a site with SSL Certificate required error:
  (This scenario is the same as the above scenario but with different discovery error type: `ERRONEOUS_SSL_CERTIFICATE`)
- Login to anitaastestwpsite2.com in `wp-admin`.
- Install, activate and connect Jetpack.
- Open the app and enter the above site address and click on `Next`.
- Click on `Login with site credentials` and enter username + password.
- Verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access your site because of a problem with the SSL Certificate. You will need to reach out to your host to resolve this.`.
- Verify that clicking on `Try again` redirects to the Username + password screen.
- Verify that clicking on `Read our troubleshooting tips` redirects to a webview from this link.
- Verify that clicking on `Help` opens the help page.
- Verify that clicking in `Sign in with Wordpress.com` redirects to the Email screen and you are able to login successfully to the app using wp.com email.

##### Scenario III: Login to a site where HTTP AUTH plugin is installed and activated.  
- Login to anitaastestwpsite.com in `wp-admin`.
- Uninstall Jetpack from the Plugins section.
- Activate `HTTP Auth` plugin from the same page. This enables HTTP Auth to the site.
- Open the app and enter the above site address and click on `Next`.
- Verify that you are redirected to the Sign in with Jetpack screen (since Jetpack is not active on the site).
- Click on Sign in in the app.
- Click on `Login with site credentials` and enter username + password.
- Verify you are redirected to the Jetpack required screen. 
- Click on Try again.
- Verify that an error snackbar is displayed stating that Jetpack is not active on the site.
- Go back to the Plugins page in `wp-admin` and deactivate `HTTP Auth` before activating and connecting Jetpack (This is because Jetpack does not support HTTP Auth plugin and you will only be able to connect Jetpack if HTTP Auth plugin is deactivated).
- Go back to the app and click on Try again.
- Verify that you are redirected back to the site credentials screen.
- Click on `Next` and verify that you are successfully redirected to the Magic link screen.

##### Scenario IV: Login to a site where the XMLRPC request is blocked.
- Login to any test site in wp-admin.
- Install the plugin `Disable XML-RPC`.
- Open the app and enter the above site address and click on Next.
- If Jetpack is not active on the site, verify that: 
     - You are redirected to the Sign in with Jetpack screen.
     - Click on Sign in in the app.
     - Click on Login with site credentials and enter username + password.
- If Jetpack is active on the site, verify that: 
     - You are redirected to the username password screen.
     - Click on Login with site credentials and enter username + password.
- In either case, verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access the XMLRPC file on your site. You will need to reach out to your host to resolve this.`.  

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

@AmandaRiu, I assigned the PR to you since you have reviewed the last two 😄 Could you trouble you for one more review please? Thank you!!

anitaa1990 committed:
Modified discovery error listener to handle the various error scenarios from the calling fragment

anitaa1990 committed:
Added logic to port over changes from the LoginSiteAddressFragment to utilise LoginBaseDiscoveryFragment

anitaa1990 committed:
Add flag to check if Jetpack is available if a discovery error takes place

anitaa1990 committed:
Added logic to verify if jetpack is installed before initiating discovery process again

anitaa1990 committed:
Merge branch 'feature/sign-in-with-self-hosted-credentials' into issue/1482-step3-discovery-error-changes

anitaa1990 committed:
Added xml + fragment for discovery error screen

anitaa1990 committed:
Added logic to redirect to discovery error screen, even if Jetpack is available

anitaa1990 committed:
Added click handlers for discovery error screen buttons

anitaa1990 committed:
Modified logic to initiate discovery from LoginBaseDiscoveryFragment

anitaa1990 committed:
Modified error message for xmlrpc errors

AmandaRiu reviewed (commented):
@anitaa1990 finished the first review round and left a code comment. I tried testing through the scenarios using our test sites page but couldn't get to the `FragmentLoginDiscoveryError` view. I tried to sign in with a WP test site that had HTTP AUTH required, but couldn't get past the first screen because `CONNECTED-SITE-INFO` was flagging it as a site that was not a WordPress site. Logging in with a site that required an SSL cert didn't give me this screen either - it would just drop me in the "Jetpack Required" screen. 

<img src="https://user-images.githubusercontent.com/5810477/67905463-a792e480-fb2e-11e9-9454-141b688ce792.png" width="300"/>

Do you have some tests sites I can use?

anitaa1990 committed:
Modified logic to check for discovery errors first before checking for Jetpack

anitaa1990 committed:
Fixed alignment issues in discovery error fragment

anitaa1990 commented on WooCommerce/src/main/res/layout/fragment_login_discovery_error.xml:
```diff
@@ -0,0 +1,261 @@
+                        android:text="@string/login_troubleshooting_tips"
+                        app:layout_constraintBottom_toBottomOf="parent"
+                        app:layout_constraintLeft_toRightOf="@+id/troubleshoot_option_icon"
+                        app:layout_constraintTop_toTopOf="parent" />
```
Thanks for the tip 😄 Done in d75b227

anitaa1990 commented:
Thanks for the quick review @AmandaRiu. 

I've invited you to two of my test sites. The invites are sent to your official email account. Do let me know if you face any issues. I've also updated the PR description with the testing steps and explanation.

PR ready for another round 👍

anitaa1990 committed:
Handle site not found discovery error which can occur even with Jetpack available

anitaa1990 committed:
Reverted logic to check for Jetpack availability during discovery errors since it can return a false negative

AmandaRiu commented:
@anitaa1990 Thank you for the test accounts and testing steps! All of them worked great except for this one: **Scenario II: Login to a site with missing rsd tags**. With that scenario I never get the "Jetpack required" view. It allows me to get all the way to the "Not a woocommerce site" view.

anitaa1990 committed:
Added logic to redirect to discovery error if xmlrpc is blocked on site

anitaa1990 commented:
Hey @amandaRiu, thanks for testing! 

> All of them worked great except for this one: Scenario II. With that scenario I never get the "Jetpack required" view. It allows me to get all the way to the "Not a woocommerce site" view.

Ah, yes! I was testing SSL certification with that test site and had to revoke the changes I made that caused this error. I forgot yo update you 🤦‍♀️  I’ve set an invalid certificate to the site now so you should be able to see the below behaviour when you try it now. The steps to reproduce this are updated in the PR as well under `Scenario II: Login to a site with SSL Certificate required error`.
<img width="300" src="https://user-images.githubusercontent.com/22608780/68001780-97444c00-fc8b-11e9-815d-5529e9d04e0c.gif">

I also faced another error when trying to login with a site that has the `Disable XMLRPC` plugin. This plugin blocks the XMLRPC calls but interestingly, the discovery API returns a successful response. But when trying to access the site details using `wp.getUserBlogs` API endpoint, results in an error. I’ve modified this behaviour in the app as well in `d58f81f ` and updated the test scenario to include this test scenario as well. To reproduce this error:
- Login to any test site in wp-admin.
- Install the plugin `Disable XML-RPC`.
- Open the app and enter the above site address and click on Next.
- If Jetpack is not active on the site, verify that: 
     - You are redirected to the Sign in with Jetpack screen.
     - Click on Sign in in the app.
     - Click on Login with site credentials and enter username + password.
- If Jetpack is active on the site, verify that: 
     - You are redirected to the username password screen.
     - Click on Login with site credentials and enter username + password.
- In either case, verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access the XMLRPC file on your site. You will need to reach out to your host to resolve this.`.

Let me know if you face any other issues! PR ready for another round 👍

AmandaRiu reviewed (approved):
Brilliant work @anitaa1990! I've run through the new test scenarios and they were all successful! 🎆  Approving this PR :shipit:
